### PR TITLE
Thin provisioning - add configuration arguments and pool commitment filter

### PR DESCRIPTION
**Pull request by tiagolobocastro**
refactor(capacity/thin): expose cli/env knobs to control overcommitment

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

feat(capacity/thin): add pool commitment filter

Allows us to configure the max pool commitment in percentage.
Example: if the commitment is 250 and the pool is 10GiB we can overcommit the pool up to 25GiB
(create 2 10GiB and 1 5GiB volume) but no further.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

feat(openapi/pool): expose pool committed bytes

This allows us to easily see if a pool is overcommitted.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

### CODE REVIEW

File: control-plane/agents/src/bin/core/main.rs [hunk 1]
```diff
@@ -88,13 +88,38 @@ pub(crate) struct CliArgs {
...
+    /// When creating pool replicas for an existing volume, each replica pool must have at least
+    /// this much free space percentage of the volume size.
+    /// Example: if this value is 40 and the pool has 40GiB free, then the max volume size allowed
+    /// to be created on the pool is 100GiB.
+    #[clap(long, env = "VOLUME_COMMITMENT_INITIAL", default_value = "40")]
+    volume_commitment_initial: u64,
+    /// When creating pool replicas for an existing volume, each replica pool must have at least
+    /// this much free space percentage of the volume size.
+    /// Example: if this value is 40, the pool has 40GiB free, then the max volume size allowed
+    /// to be created on the pool is 100GiB.
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/main.rs]: Q: Do these parameters represent the same thing? The comments are identical.

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/agents/src/bin/core/main.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/main.rs]: Sigh... they're identical because I messed up the rebase... I'll fix that up next week.

**Review by blaisedias — APPROVED** [review]
Looks plausible to me.

File: control-plane/agents/src/bin/core/main.rs [hunk 2]
```diff
@@ -88,13 +88,42 @@ pub(crate) struct CliArgs {
...
 
+/// Cluster wide thin provisioning parameters.
+#[derive(Debug, clap::Parser, Clone)]
+pub(crate) struct ThinArgs {
+    /// The allowed pool commitment limit when dealing with thin provisioned volumes.
+    /// Example: If the commitment is 250 and the pool is 10GiB we can overcommit the pool
+    /// up to 25GiB (create 2 10GiB and 1 5GiB volume) but no further.
+    #[clap(long, env = "POOL_COMMITMENT_%", value_parser = value_parse_percent, default_value = "250%")]
+    pool_commitment: u64,
+    /// When creating pool replicas for a new volume, each replica pool must have at least this
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/main.rs]: `When creating pool replicas for a new volume` -> `When creating replicas for a new volume` ?

File: control-plane/agents/src/bin/core/main.rs [hunk 3]
```diff
@@ -88,13 +88,42 @@ pub(crate) struct CliArgs {
...
+    /// much free space percentage of the volume size.
+    /// Example: if this value is 40 and the pool has 40GiB free, then the max volume size allowed
+    /// to be created on the pool is 100GiB.
+    #[clap(long, env = "VOLUME_COMMITMENT_%_INITIAL", value_parser = value_parse_percent, default_value = "40%")]
+    volume_commitment_initial: u64,
+    /// When creating pool replicas for an existing volume, each replica pool must have at least
+    /// this much free space percentage of the volume size.
+    /// Example: if this value is 40, the pool has 40GiB free, then the max volume size allowed
+    /// to be created on the pool is 100GiB.
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/main.rs:112]: The doc comment still look identical? Should we mention how one differs from other?

File: control-plane/agents/src/bin/core/pool/registry.rs [hunk 4]
```diff
@@ -64,31 +63,42 @@ impl Registry {
...
-        Err(PoolNotFound { pool_id })
+        Err(PoolNotFound {
+            pool_id: pool_id.to_owned(),
+        })
     }
 
-    /// Get all pools
-    pub(crate) async fn pool_states_inner(&self) -> Vec<PoolState> {
+    /// Get all controller pools states.
+    pub(crate) async fn ctrl_pool_states(&self) -> Vec<CtrlPoolState> {
```
Comment (Abhinandan-Purkait) [inline control-plane/agents/src/bin/core/pool/registry.rs:79]: What do we mean by controller pool state?

File: control-plane/agents/src/bin/core/main.rs (hunk 3 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/main.rs:112]: One is for new, one is not existing volumes, that's the difference.
Yes that's a good idea, I'll say how they differ.

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**